grader_utils/__init__.py

```diff
@@ -0,0 +1,196 @@
+
+        # Fields that will be outputted later
+        self.killed = False
+        self.stdout_res = None
+        self.stderr_res = None
+        # If the process timeouts
+        self.exit_timeout = False
+        # If the process is killed by SIGKILL or exit normally
+        self.exit_force = False
+        # If
+        # - the process timeouts,
+        # - or any of the processes in the process group timeouts while main
+        #   process exited.
+        # Either case we will try to kill the process group using SIGTERM.
+        self.exit_timeout_pg = False
+        # If
+        # - the process is not killed by SIGTERM
+        # - or any of the processes in the process group is not killed by
+        #   SIGTERM.
+        # In either case we will try to kill the process group using SIGKILL.
+        self.exit_force_pg = False
+
+    def run(self):
+        """Run the command as a separate process and return immediately.
+
+        Call join() later to get result of process. Timeout does not start
+        until join() is called. Additionally send the delayed_inputs and inputs
+        to process.
+        """
+        def target():
+            # Pass these inputs to STDIN with delays
+            for i in self.delayed_inputs:
+                if type(i) is int or type(i) is float:
+                    time.sleep(i)
+                elif type(i) is bytes:
+                    try:
+                        self.process.stdin.write(i) 
+                    except IOError as e:
+                        lg.info(
+                            "Input: {} failed to write to stdin due to\n{}".format(i, e)
+                            )
+                        break
+            if self.disable_communicate:
+                self.process.wait()
+            else:
+                self.stdout_res, self.stderr_res = self.process.communicate(
+                    input=self.inputs)
+
+        try:
+            self.process = Popen(self.command, stdin=self.stdin,
+                                 stdout=self.stdout, stderr=self.stderr,
+                                 start_new_session=True, cwd=self.cwd, env=self.env)
+        except OSError:
+            lg.error("Couldn't Popen command {}".format(self.command))
+            raise
+        self.thread = Thread(target=target)
+        self.thread.start()
+
+    def join(self):
+        """Join with the thread and get the result of the separate process.
+
+        Will timeout based on self.timeout. Timeout involves sending SIGTERM
+        followed by SIGKILL if that is ignored.
+        """
+        self.thread.join(timeout=self.timeout)
+        alive = self.is_alive if self.check_pg_alive else self.thread.is_alive
+        self.exit_timeout = self.thread.is_alive()
+        if alive():
+            self.exit_timeout_pg = self.check_pg_alive
+            lg.info("{}: is still alive. Terminating now...".format(self.process.pid))
+            os.killpg(self.process.pid, signal.SIGTERM)
```

Comment:
This can't be done with the Popen object?

Reply:
Apparently not